File: Python/Euler022.py
#!/usr/bin/env python3
from __future__ import print_function
import re


BASE_LETTER_SCORE = ord('a') - 1


def main():
    names = sorted(read_file('input/names.txt'))
    scores = (score_name(name) * idx for idx, name in enumerate(names, 1))
    final_score = sum(scores)
    print(final_score)


def read_file(path):
    with open(path) as f:
        return (re.sub(r'\W', '', name) for name in f.read().split(','))


def score_name(name):
    return sum(score_letter(c) for c in name.lower())


def score_letter(c):
    return ord(c) - BASE_LETTER_SCORE

if __name__ == '__main__':
    main()

File: Ruby/Euler022.rb
#!/usr/bin/env ruby

BASE_LETTER_SCORE = 'a'.ord - 1

def main()
  names = load_file 'input/names.txt'
  names = names.sort
  scores = names.each_with_index.map { |n, idx| score_name(n) * (idx + 1) }
  final_score = scores.sum
  puts final_score
end

def load_file(path)
  File.open(path) do |f|
    f.read.split(',').map { |name| name.gsub(/\W/, '') }
  end
end

def score_name(name)
  return name.downcase.chars.map { |c| score_letter c }.sum
end

def score_letter(c)
  return c.ord - BASE_LETTER_SCORE
end

if __FILE__ == $0
  main
end
